bee-tangle/src/vertices.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2021 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::vertex::Vertex;
+
+use bee_message::MessageId;
+
+use hashbrown::{hash_map::DefaultHashBuilder, raw::RawTable};
+use rand::Rng;
+use tokio::sync::{RwLock, RwLockMappedWriteGuard, RwLockReadGuard, RwLockWriteGuard};
+
+use std::{
+    hash::{BuildHasher, Hash, Hasher},
+    num::NonZeroUsize,
+    sync::atomic::{AtomicUsize, Ordering},
+};
+
+fn equivalent_id(message_id: &MessageId) -> impl Fn(&(MessageId, Vertex)) -> bool + '_ {
+    move |(k, _)| message_id.eq(k)
+}
+
+type Table = RwLock<RawTable<(MessageId, Vertex)>>;
+
+pub(crate) struct Vertices {
+    hash_builder: DefaultHashBuilder,
+    tables: Box<[Table]>,
+    len: AtomicUsize,
+}
+
+impl Vertices {
+    pub(crate) fn new(num_partitions: NonZeroUsize) -> Self {
+        Self {
+            hash_builder: DefaultHashBuilder::default(),
+            tables: (0..num_partitions.into())
+                .map(|_| RwLock::new(RawTable::default()))
+                .collect(),
+            len: AtomicUsize::default(),
+        }
+    }
+
+    fn make_hash(&self, message_id: &MessageId) -> u64 {
+        let mut state = self.hash_builder.build_hasher();
+        message_id.hash(&mut state);
+        state.finish()
+    }
+
+    fn make_hasher(&self) -> impl Fn(&(MessageId, Vertex)) -> u64 + '_ {
+        move |(message_id, _)| self.make_hash(message_id)
+    }
+
+    fn get_table(&self, hash: u64) -> &Table {
+        let index = hash as usize % self.tables.len();
+        // SAFETY: `index < self.tables.len()` by construction.
+        unsafe { self.tables.get_unchecked(index) }
+    }
+
+    pub(crate) async fn get(&self, message_id: &MessageId) -> Option<RwLockReadGuard<'_, Vertex>> {
+        let hash = self.make_hash(message_id);
+        let table = self.get_table(hash).read().await;
+
+        RwLockReadGuard::try_map(table, |table| {
+            table.get(hash, equivalent_id(message_id)).map(|(_, v)| v)
+        })
+        .ok()
+    }
+
+    pub(crate) async fn get_mut(&self, message_id: &MessageId) -> Option<RwLockMappedWriteGuard<'_, Vertex>> {
+        let hash = self.make_hash(message_id);
+        let table = self.get_table(hash).write().await;
+
+        RwLockWriteGuard::try_map(table, |table| {
+            table.get_mut(hash, equivalent_id(message_id)).map(|(_, v)| v)
+        })
+        .ok()
+    }
+
+    pub(crate) async fn pop_random(&self) -> Option<Vertex> {
+        let index = rand::thread_rng().gen_range(0..self.tables.len());
+        // SAFETY: `index < self.tables.len()` by construction.
+        let mut table = unsafe { self.tables.get_unchecked(index) }.write().await;
```

Comment:
Would this work here to get rid of the `unsafe`?

```suggestion
        let mut table = self.tables.choose(&mut rand::thread_rng()).write().await;
```

Reply:
I believe the unchecked get was here for performance reason as we know the index is valid. @pvdrz ?

Reply:
Ah that makes sense, I think performance would definitely be better that way.

Reply:
yes, but I wouldn't be surprised if rust can optimize it away, who knows